File: util/unit/UnitLevel.java
package common.util.unit;

import common.io.InStream;
import common.io.json.JsonClass;
import common.io.json.JsonField;
import common.pack.Identifier;
import common.pack.IndexContainer.IndexCont;
import common.pack.IndexContainer.Indexable;
import common.pack.PackData;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

@JsonClass
@JsonClass.JCGeneric(Identifier.class)
@IndexCont(PackData.class)
public class UnitLevel implements Indexable<PackData, UnitLevel> {

	@JsonField
	public int[] lvs = new int[20];

	public final List<Unit> units = new ArrayList<>();

	@JsonField
	@JsonClass.JCIdentifier
	public Identifier<UnitLevel> id;

	public UnitLevel(Identifier<UnitLevel> ID, InStream is) {
		id = ID;
		zread(is);
	}

	public UnitLevel(Identifier<UnitLevel> ID, UnitLevel ul) {
		id = ID;
		lvs = ul.lvs.clone();
	}

	public UnitLevel(int[] inp) {
		lvs = Arrays.copyOf(inp, 20);
	}

	@JsonClass.JCConstructor
	public UnitLevel() {}

	@Override
	public boolean equals(Object o) {
		if (this == o)
			return true;
		if (!(o instanceof UnitLevel))
			return false;
		UnitLevel ul = (UnitLevel) o;
		if (lvs.length != ul.lvs.length)
			return false;
		for (int i = 0; i < lvs.length; i++)
			if (lvs[i] != ul.lvs[i])
				return false;
		return id == null || ul.id == null || id.equals(ul.id);
	}

	@Override
	public Identifier<UnitLevel> getID() {
		return id;
	}

	public float getMult(int lv) {
		int dec = lv;
		float d = 0.8f;

		for (int mul : lvs) {
			if (dec >= 10) {
				d += mul * 0.1f;
				dec -= 10;
			} else {
				d += mul * dec * 0.01f;
				break;
			}
		}
		return d;
	}

	@Override
	public String toString() {
		StringBuilder ans = new StringBuilder("{");
		for (int set : lvs) {
			if (ans.length() > 1)
				ans.append(", ");
			ans.append(set);
		}
		ans.append("}");
		return ans.toString();
	}

	private void zread(InStream is) {
		int ver = is.nextInt();
		int[] lvs = new int[3];
		if (ver == 1) {
			int[] vs = is.nextIntsB();
			lvs[0] = vs[0];
			lvs[1] = vs[1];
			lvs[2] = vs[2];
		} else {
			int[][] vs = is.nextIntsBB();
			lvs[0] = vs[1][0];
			lvs[1] = vs[2][0];
			lvs[2] = vs[3][0];
		}

		int pre = 0, mul = 20;
		for (int i = 0; i < 3; i++) {
			for (int j = pre; j < lvs[i]; j++)
				if(j < 20)
					this.lvs[j] = mul;
			mul /= 2;
			pre = lvs[i];
		}
	}

}

File: util/unit/UnitLevel.kt
package common.util.unit

import common.io.InStream
import common.pack.IndexContainer.IndexCont
import common.pack.IndexContainer.Indexable
import common.pack.PackData
import common.util.unit.Unitimport
import java.util.*

com.google.api.client.json.jackson2.JacksonFactory
@IndexCont(PackData::class)
class UnitLevel : Indexable<PackData?, UnitLevel?> {
    val lvs = IntArray(3)
    val units: MutableList<Unit> = ArrayList()
    var id: PackData.Identifier<UnitLevel>? = null

    constructor(ID: PackData.Identifier<UnitLevel>?, `is`: InStream) {
        id = ID
        zread(`is`)
    }

    constructor(ID: PackData.Identifier<UnitLevel>?, ul: UnitLevel) {
        id = ID
        for (i in 0..2) lvs[i] = ul.lvs[i]
    }

    constructor(inp: IntArray) {
        var `val` = -1
        for (i in inp.indices) {
            if (`val` != inp[i]) {
                `val` = inp[i]
                if (`val` == 10) lvs[0] = i
                if (`val` == 5) lvs[1] = i
                if (`val` == 0) lvs[2] = i
            }
        }
        if (lvs[1] == 0) lvs[1] = inp.size
        if (lvs[2] == 0) lvs[2] = inp.size
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o !is UnitLevel) return false
        val ul = o
        if (lvs.size != ul.lvs.size) return false
        for (i in lvs.indices) if (lvs[i] != ul.lvs[i]) return false
        return id == null || ul.id == null || id!!.equals(ul.id)
    }

    override fun getID(): PackData.Identifier<UnitLevel>? {
        return id
    }

    fun getMult(lv: Int): Double {
        var dec = lv
        var pre = 0
        var mul = 20
        var d = 0.8
        for (i in lvs.indices) {
            val dur = lvs[i] - pre
            if (dec > dur * 10) {
                d += mul * dur * 0.1
                dec -= dur * 10
            } else {
                d += mul * dec * 0.01
                break
            }
            mul /= 2
            pre = lvs[i]
        }
        return d
    }

    override fun toString(): String {
        var ans = "{"
        for (set in lvs) {
            if (ans.length > 1) ans += ", "
            ans += set
        }
        ans += "}"
        return ans
    }

    private fun zread(`is`: InStream) {
        val ver: Int = `is`.nextInt()
        if (ver == 1) {
            val vs: IntArray = `is`.nextIntsB()
            lvs[0] = vs[0]
            lvs[1] = vs[1]
            lvs[2] = vs[2]
        } else {
            val vs: Array<IntArray> = `is`.nextIntsBB()
            lvs[0] = vs[1][0]
            lvs[1] = vs[2][0]
            lvs[2] = vs[3][0]
        }
    }

    companion object {
        var def: UnitLevel? = null
    }
}
